### PR TITLE
feat: pass ip details for factorsAI

### PR DESCRIPTION
**Pull request by aashishmalik**
## What are the changes introduced in this PR?

We want to pass ip details for factorsAI.

Write a brief explainer on your code changes.

Added a function that will populate ip details , if in `context we don't get IP details` then we will populate this from `message.request_ip.`

## Please explain the objectives of your changes below

Put down any required details on the broader aspect of your changes. If there are any dependent changes, **mandatorily** mention them here

### Type of change

If the pull request is a **bug-fix**, **enhancement** or a **refactor**, please fill in the details on the changes made.

- Existing capabilities/behavior

- New capabilities/behavior

If the pull request is a **new feature**,

### Any technical or performance related pointers to consider with the change?

N/A

### Any new dependencies introduced with this change?

N/A

### Any new generic utility introduced or modified. Please explain the changes.

N/A

### If the PR has changes in more than 10 files, please mention why the changes were not split into multiple PRs.

N/A

### If multiple linear tasks are associated with the PR changes, please elaborate on the reason:

N/A

<hr>

### Developer checklist

- [ ] **No breaking changes are being introduced.**

- [ ] Are all related docs linked with the PR?

- [ ] Are all changes manually tested?

- [ ] Does this change require any documentation changes?

- [ ] Are relevant unit and component test-cases added?

### Reviewer checklist

- [ ] Is the type of change in the PR title appropriate as per the changes?

- [ ] Verified that there are no credentials or confidential data exposed with the changes.

### CODE REVIEW

File: src/v0/destinations/factorsai/transform.js [hunk 1]
```diff
@@ -13,6 +13,15 @@ const { JSON_MIME_TYPE } = require('../../util/constant');
 
 const { ConfigCategories, mappingConfig, BASE_URL } = require('./config');
 
+function populateIpDetails(requestJson, message) {
+  const payload = requestJson;
+  if (message.context && requestJson.context) {
```
Comment (sanpj2292) [inline src/v0/destinations/factorsai/transform.js:18]: I don't feel this condition is right. May be this `payload.context && (message.context || message.request_ip)` will work better for us

But even below way of rewriting the logic might be good for us
```javascript
if (message.context || message.request_ip) {
  // assuming `payload.context` is an object
  payload.context = { ...(payload.context || {}), ip: message.context?.ip || message.request_ip }
}
```
Please do evaluate this before making changes

**Comment by sanpj2292** [comment]
Just to confirm, are we ok with sending IP information in all message types(identify, track, page, group etc,.) ? @aashishmalik

File: src/v0/destinations/factorsai/transform.js (hunk 1 above)
Comment (aashishmalik) [reply src/v0/destinations/factorsai/transform.js]: updated, added more test case

**Comment by aashishmalik** [comment]
> Just to confirm, are we ok with sending IP information in all message types(identify, track, page, group etc,.) ? @aashishmalik

got this response from factorsAI "would be good to get the ip on all requests"

File: src/v0/destinations/factorsai/transform.js (hunk 1 above)
Comment (sanpj2292) [reply src/v0/destinations/factorsai/transform.js]: I think we should consider the condition where `payload.context` can be anything other than object. Apart from that I'm good with the changes

Comment (sanpj2292) [reply src/v0/destinations/factorsai/transform.js]: context according to rudderstack spec is an object. Hence closing this comment

**Review by sanpj2292 — APPROVED** [review]

File: src/v0/destinations/factorsai/transform.js (hunk 1 above)
Comment (aashishmalik) [reply src/v0/destinations/factorsai/transform.js]: it must be an object as per rudder contract

**Review by mihir-4116 — APPROVED** [review]

**Pull request merged**